src/handler/lib/policy.c

```diff
@@ -1285,25 +1286,127 @@ __attribute__((noinline)) int get_wallet_internal_script_hash(
 
 #pragma GCC diagnostic pop
 
-int get_policy_address_type(const policy_node_t *policy) {
-    // legacy, native segwit, wrapped segwit, or taproot
-    switch (policy->type) {
+// For a standard descriptor template, return the corresponding BIP44 purpose
+// Otherwise, returns -1.
+static int get_bip44_purpose(const policy_node_t *descriptor_template) {
+    const policy_node_key_placeholder_t *kp = NULL;
+    int purpose = -1;
+    switch (descriptor_template->type) {
         case TOKEN_PKH:
-            return ADDRESS_TYPE_LEGACY;
+            kp = ((const policy_node_with_key_t *) descriptor_template)->key_placeholder;
+            purpose = 44;  // legacy
+            break;
         case TOKEN_WPKH:
-            return ADDRESS_TYPE_WIT;
-        case TOKEN_SH:
-            // wrapped segwit
-            if (resolve_node_ptr(&((const policy_node_with_script_t *) policy)->script)->type ==
-                TOKEN_WPKH) {
-                return ADDRESS_TYPE_SH_WIT;
+            kp = ((const policy_node_with_key_t *) descriptor_template)->key_placeholder;
+            purpose = 84;  // native segwit
+            break;
+        case TOKEN_SH: {
+            const policy_node_t *inner = resolve_node_ptr(
+                &((const policy_node_with_script_t *) descriptor_template)->script);
+            if (inner->type != TOKEN_WPKH) {
+                return false;
             }
-            return -1;
+
+            kp = ((const policy_node_with_key_t *) inner)->key_placeholder;
+            purpose = 49;  // nested segwit
+            break;
+        }
         case TOKEN_TR:
-            return ADDRESS_TYPE_TR;
+            if (((const policy_node_tr_t *) descriptor_template)->tree != NULL) {
+                return false;
+            }
+
+            kp = ((const policy_node_tr_t *) descriptor_template)->key_placeholder;
+            purpose = 86;  // standard single-key P2TR
+            break;
         default:
-            return -1;
+            return false;
```

Comment:
Looks a bit weird to me to have a `static int` function returning `false`

Reply:
Oops, good catch, the would return 0, which is a bug as that could technically be a valid purpose.
Remnant of a previous function returning `bool` (but how does that _not_ give a warning? Aaaah, C..)!

Reply:
Fixed (and rebased).